elastic/kibana
# [Alerting] Update rules detail page to resolve SO IDs if necessary

Resolve https://github.com/elastic/kibana/issues/108074

## Summary

This PR:
* Adds an internal `resolve` HTTP API that calls the saved objects resolve function
* Updates the rule details page to handle 404 Not found errors from the existing GET API and try to resolve those rule IDs using the new internal `resolve` API. Handling for non-404 errors is unchanged.
* Uses the provided `spacesOss` UI components to
  * redirect to new URL and show toaster message when resolve outcome is `aliasMatch`
  * show warning callout when resolve outcome is `conflict`

## To Verify
* Run 7.14 using `yarn es snapshot --ssl -E path.data=../data` and create some rules, including rules in a non-default space
  * Bookmark or copy the rule details page URLs
    * https://localhost:5601/app/management/insightsAndAlerting/triggersActions/rule/{ruleId}
    * https://localhost:5601/s/new-space/app/management/insightsAndAlerting/triggersActions/rule/{anotherRuleId}
* Switch to this branch and make the following changes: ([see Chris's POC PR for example](https://github.com/elastic/kibana/pull/107611/files))
  * Change the alerting SO to `multiple-isolated` and set `convertToMultiNamespaceTypeVersion: '8.0.0' in `x-pack/plugins/alerting/server/saved_objects/index.ts`
  * Add a dummy migration to `x-pack/plugins/alerting/server/saved_objects/migrations.ts`

```
const migrationActions800 = createEsoMigration(
    encryptedSavedObjects,
    (doc: SavedObjectUnsanitizedDoc<RawAlert>): doc is SavedObjectUnsanitizedDoc<RawAlert> => true,
    (doc) => doc // no-op
  );

'8.0.0': executeMigrationWithErrorHandling(migrationActions800, '8.0.0'),
```

* Run this branch using the same data from 7.14 
* Go to the bookmarked URLs from before and make sure your network tab is open
* For rules in the default space, you should see a successful GET request when you go to the bookmarked URL
* For rules in the non-default space, you should see a 404 for the GET and then a successful `resolve` request when you go to the bookmarked URL. You should be redirected to the rule details page for the updated rule ID and you should see a toaster pop up.
* Not sure how to recreate a `conflict` outcome but if you want to see the `conflict` callout, you could modify `x-pack/plugins/triggers_actions_ui/public/application/sections/alert_details/components/alert_details_route.tsx` to set the outcome to `conflict` after the resolve call.

### Checklist
